verifier/verifier.go

```diff
@@ -662,51 +678,210 @@ func verifyExpiry(outcome *notation.VerificationOutcome) *notation.ValidationRes
+				Type:   trustpolicy.TypeAuthenticTimestamp,
+				Action: outcome.VerificationLevel.Enforcement[trustpolicy.TypeAuthenticTimestamp],
+			}
+		}
+		tsaCertChain, err := signedToken.Verify(ctx, x509.VerifyOptions{
+			CurrentTime: timestamp.Value,
+			Roots:       rootCertPool,
+		})
+		if err != nil {
+			return &notation.ValidationResult{
+				Error:  fmt.Errorf("failed to verify the timestamp countersignature with error: %w", err),
+				Type:   trustpolicy.TypeAuthenticTimestamp,
+				Action: outcome.VerificationLevel.Enforcement[trustpolicy.TypeAuthenticTimestamp],
+			}
+		}
+		// 3. Validate timestamping certificate chain
+		logger.Info("Validating timestamping certificate chain...")
+		if err := nx509.ValidateTimestampingCertChain(tsaCertChain); err != nil {
+			return &notation.ValidationResult{
+				Error:  fmt.Errorf("failed to validate the timestamping certificate chain with error: %w", err),
+				Type:   trustpolicy.TypeAuthenticTimestamp,
+				Action: outcome.VerificationLevel.Enforcement[trustpolicy.TypeAuthenticTimestamp],
+			}
+		}
+		logger.Info("TSA identity is: ", tsaCertChain[0].Subject)
+		// 4. Perform the timestamping certificate chain revocation check
+		if !signatureVerification.SkipTimestampRevocationCheck {
+			logger.Info("Checking timestamping certificate chain revocation...")
+			certResults, err := r.Validate(tsaCertChain, timestamp.Value)
+			if err != nil {
+				return &notation.ValidationResult{
+					Error:  fmt.Errorf("failed to check timestamping certificate chain revocation with error: %w", err),
+					Type:   trustpolicy.TypeAuthenticTimestamp,
+					Action: outcome.VerificationLevel.Enforcement[trustpolicy.TypeAuthenticTimestamp],
+				}
+			}
+			finalResult, problematicCertSubject := revocationFinalResult(certResults, tsaCertChain, logger)
+			switch finalResult {
+			case revocationresult.ResultOK:
+				logger.Debug("No verification impacting errors encountered while checking timestamping certificate chain revocation, status is OK")
+			case revocationresult.ResultRevoked:
+				return &notation.ValidationResult{
+					Error:  fmt.Errorf("timestamping certificate with subject %q is revoked", problematicCertSubject),
+					Type:   trustpolicy.TypeAuthenticTimestamp,
+					Action: outcome.VerificationLevel.Enforcement[trustpolicy.TypeAuthenticTimestamp],
+				}
+			default:
+				// revocationresult.ResultUnknown
+				return &notation.ValidationResult{
+					Error:  fmt.Errorf("timestamping certificate with subject %q revocation status is unknown", problematicCertSubject),
+					Type:   trustpolicy.TypeAuthenticTimestamp,
+					Action: outcome.VerificationLevel.Enforcement[trustpolicy.TypeAuthenticTimestamp],
+				}
+			}
+		}
+		// 5. Check the timestamp against the signing certificate chain
+		logger.Info("Checking the timestamp against the signing certificate chain...")
+		logger.Infof("Timestamp range: [%v, %v]", timestamp.Value.Add(-timestamp.Accuracy), timestamp.Value.Add(timestamp.Accuracy))
```

Comment:
Looks like the `Timestamp` struct needs a `String()` method.

Reply:
Yeah, that's true. Since it's not a blocker, we could improve it in a future iteration.

Reply:
Can you create an issue for tracking?

Reply:
Issue created: https://github.com/notaryproject/tspclient-go/issues/29